File: tests/scalers/openstack-swift.test.ts
import * as sh from 'shelljs'
import * as tmp from 'tmp'
import * as fs from 'fs'
import test from 'ava';

import SwiftClient from './openstack/openstack-swift-helper';

var swiftClient: SwiftClient

const openstackSwiftURL = process.env['OS_SWIFT_URL']
const openstackUserID = process.env['OS_USER_ID']
const openstackPassword = process.env['OS_PASSWORD']
const openstackProjectID = process.env['OS_PROJECT_ID']
const openstackAuthURL = process.env['OS_AUTH_URL']
const openstackRegionName = process.env['OS_REGION_NAME']

const testNamespace = 'openstack-swift-password-test'

const swiftContainerName = 'my-container-password-test'
const swiftContainerObjects = [
  '1/',
  '1/2/',
  '1/2/3/',
  '1/2/3/4/',
  '1/2/3/hello-world.txt',
  '1/2/hello-world.txt',
  '1/hello-world.txt',
  '2/',
  '2/hello-world.txt',
  '3/',
]

const deploymentName = 'hello-node-password'
const deploymentImage = "k8s.gcr.io/echoserver:1.4"

const secretYamlFile = tmp.fileSync()
const scaledObjectYamlFile = tmp.fileSync()
const triggerAuthenticationYamlFile = tmp.fileSync()

test.before(t => {
  sh.config.silent = true

  if (!openstackUserID) {
    t.fail('OS_USER_ID environment variable is required for running tests')
  }

  if (!openstackPassword) {
    t.fail('OS_PASSWORD environment variable is required for running tests')
  }

  if (!openstackProjectID) {
    t.fail('OS_PROJECT_ID environment variable is required for running tests')
  }

  if (!openstackAuthURL) {
    t.fail('OS_AUTH_URL environment variable is required for running tests')
  }
});

test.serial.before(async t => {
  try {
    swiftClient = await SwiftClient.create()

    if (swiftClient) {
      await swiftClient.createContainer(swiftContainerName)

      for (const object of swiftContainerObjects) {
        await swiftClient.createObject(swiftContainerName, object)
      }
    }
  } catch (err) {
    t.fail(err.message)
  }
});

test.serial.before(async t => {
  const base64OpenstackUserID = Buffer.from(openstackUserID).toString('base64')
  const base64OpenstackPassword = Buffer.from(openstackPassword).toString('base64')
  const base64OpenstackProjectID = Buffer.from(openstackProjectID).toString('base64')
  const base64OpenstackAuthURL = Buffer.from(openstackAuthURL).toString('base64')
  var base64OpenstackRegionName = ""

  if (openstackRegionName) base64OpenstackRegionName = Buffer.from(openstackRegionName).toString('base64')

  fs.writeFileSync(secretYamlFile.name, swiftSecretYaml
    .replace('{{OS_USER_ID}}', base64OpenstackUserID)
    .replace('{{OS_PASSWORD}}', base64OpenstackPassword)
    .replace('{{OS_PROJECT_ID}}', base64OpenstackProjectID)
    .replace('{{OS_AUTH_URL}}', base64OpenstackAuthURL)
    .replace('{{OS_REGION_NAME}}', base64OpenstackRegionName)
  )

  fs.writeFileSync(triggerAuthenticationYamlFile.name, swiftTriggerAuthenticationYaml)

  fs.writeFileSync(scaledObjectYamlFile.name, swiftScaledObjectYaml
    .replace('{{DEPLOYMENT_NAME}}', deploymentName)
    .replace('{{OS_SWIFT_URL}}', openstackSwiftURL)
    .replace('{{CONTAINER_NAME}}', swiftContainerName)
  )

  sh.exec(`kubectl create namespace ${testNamespace}`)

  t.is(
    0,
    sh.exec(`kubectl create deployment ${deploymentName} --image=${deploymentImage} --namespace ${testNamespace}`).code,
    'Creating deployment for scaling tests should work.'
  )

  t.is(
    0,
    sh.exec(`kubectl apply -f ${secretYamlFile.name} --namespace ${testNamespace}`).code,
    'Creating secret for storing OpenStack credentials should work.'
  )

  t.is(
    0,
    sh.exec(`kubectl apply -f ${triggerAuthenticationYamlFile.name} --namespace ${testNamespace}`).code,
    'Creating triggerAuthentication should work.'
  )
});

test.serial('Total number of objects inside container should be 10', async t => {
  try {
    const objectCount = await swiftClient.getObjectCount(swiftContainerName)
    t.is(objectCount, 10);
  } catch (err) {
    t.fail(err.message)
  }
});

test.serial('Deployment should have 1 replica on start', t => {
  let replicaCount = ''

  for (let i = 0; i < 20 && replicaCount !== '1'; i++) {
    replicaCount = sh.exec(
        `kubectl get deployment.apps/${deploymentName} --namespace ${testNamespace} -o jsonpath="{.spec.replicas}"`
    ).stdout

    if (replicaCount !== '1') {
        sh.exec('sleep 1s')
    }
  }

  t.is(
    replicaCount,
    '1',
    'Replica count should start out as 1'
  )
})

test.serial('Deployment should be scaled to 10 after creating ScaledObject', t => {
  t.is(
    0,
    sh.exec(`kubectl apply -f ${scaledObjectYamlFile.name} --namespace ${testNamespace}`).code,
    'Creating scaledObject should work.'
  )

  let replicaCount = ''

  for (let i = 0; i < 40 && replicaCount !== '10'; i++) {
    replicaCount = sh.exec(
        `kubectl get deployment.apps/${deploymentName} --namespace ${testNamespace} -o jsonpath="{.spec.replicas}"`
    ).stdout

    if (replicaCount !== '10') {
        sh.exec('sleep 3s')
    }
  }

  t.is(
    replicaCount,
    '10',
    'Replica count should be 10 after creating ScaledObject'
  )
})

test.serial('Deployment should be scaled to 5 after deleting 5 objects in container', async t => {
  try {
    let replicaCount = ''

    await swiftClient.deleteObject(swiftContainerName, '1/2/hello-world.txt')
    await swiftClient.deleteObject(swiftContainerName, '1/hello-world.txt')
    await swiftClient.deleteObject(swiftContainerName, '2/')
    await swiftClient.deleteObject(swiftContainerName, '2/hello-world.txt')
    await swiftClient.deleteObject(swiftContainerName, '3/')

    for (let i = 0; i < 110 && replicaCount !== '5'; i++) {
      replicaCount = sh.exec(
          `kubectl get deployment.apps/${deploymentName} --namespace ${testNamespace} -o jsonpath="{.spec.replicas}"`
      ).stdout

      if (replicaCount !== '5') {
          sh.exec('sleep 3s')
      }
    }

    t.is(
      replicaCount,
      '5',
      'Replica count should be 5 after creating deleting 5 objects in conatainer'
    )
  } catch (err) {
    t.fail(err.message)
  }
})

test.serial('Deployment should be scaled to 0 after deleting all objects in container', async t => {
  try {
    let replicaCount = '10'

    const { isEmpty, response } = await swiftClient.deleteAllObjects(swiftContainerName)

    if(!isEmpty) {
      t.fail(`Could not delete all objects inside container to test scaling to zero. Swift API returned: ${response}`)
    }

    for (let i = 0; i < 20 && replicaCount !== '0'; i++) {
      replicaCount = sh.exec(
          `kubectl get deployment.apps/${deploymentName} --namespace ${testNamespace} -o jsonpath="{.spec.replicas}"`
      ).stdout

      if (replicaCount !== '0') {
          sh.exec('sleep 3s')
      }
    }

    t.is(
      replicaCount,
      '0',
      'Replica count should be 0 after creating ScaledObject'
    )
  } catch (err) {
    t.fail(err.message)
  }
})

test.after.always('Clean up OpenStack Swift container', async t => {
  try {
    if(swiftClient) {
      await swiftClient.deleteContainer(swiftContainerName)
    }
  } catch (err) {
    t.fail(err.message)
  }
});

test.after.always.cb('Clean up Secret, Deployment and openstack-swift scaler resources', t => {
  const resources = [
    'scaledobject.keda.sh/swift-password-scaledobject',
    'triggerauthentication.keda.sh/keda-trigger-password-openstack-secret',
    'secret/openstack-password-secrets',
    `deployment.apps/${deploymentName}`,
  ]

  for (const resource of resources) {
    sh.exec(`kubectl delete ${resource} --namespace ${testNamespace}`)
  }

  sh.exec(`kubectl delete namespace ${testNamespace}`)

  t.end();
});

const swiftSecretYaml = `
apiVersion: v1
kind: Secret
metadata:
  name: openstack-password-secrets
type: Opaque
data:
  userID: {{OS_USER_ID}}
  password: {{OS_PASSWORD}}
  projectID: {{OS_PROJECT_ID}}
  authURL: {{OS_AUTH_URL}}
`

const swiftTriggerAuthenticationYaml = `
apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: keda-trigger-password-openstack-secret
spec:
  secretTargetRef:
  - parameter: userID
    name: openstack-password-secrets
    key: userID
  - parameter: password
    name: openstack-password-secrets
    key: password
  - parameter: projectID
    name: openstack-password-secrets
    key: projectID
  - parameter: authURL
    name: openstack-password-secrets
    key: authURL
`

const swiftScaledObjectYaml = `
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: swift-password-scaledobject
spec:
  scaleTargetRef:
    name: {{DEPLOYMENT_NAME}}
  pollingInterval: 10
  cooldownPeriod: 10
  minReplicaCount: 0
  triggers:
  - type: openstack-swift
    metadata:
      containerName: {{CONTAINER_NAME}}
      objectCount: '1'
    authenticationRef:
        name: keda-trigger-password-openstack-secret
`

File: tests/scalers/openstack_swift/openstack_swift_test.go
//go:build e2e
// +build e2e

package openstack_test

import (
	"encoding/base64"
	"fmt"
	"os"
	"testing"

	"github.com/gophercloud/gophercloud"
	"github.com/joho/godotenv"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"k8s.io/client-go/kubernetes"

	. "github.com/kedacore/keda/v2/tests/helper"
	helper "github.com/kedacore/keda/v2/tests/scalers/openstack_swift"
)

// Load environment variables from .env file
var _ = godotenv.Load("../../.env")

const (
	testName = "openstack-swift-test"
)

var (
	testNamespace             = fmt.Sprintf("%s-ns", testName)
	deploymentName            = fmt.Sprintf("%s-deployment", testName)
	secretName                = fmt.Sprintf("%s-secret", testName)
	triggerAuthenticationName = fmt.Sprintf("%s-ta", testName)
	scaledObjectName          = fmt.Sprintf("%s-so", testName)
	userID                    = os.Getenv("OPENSTACK_USER_ID")
	password                  = os.Getenv("OPENSTACK_PASSWORD")
	projectID                 = os.Getenv("OPENSTACK_PROJECT_ID")
	authURL                   = os.Getenv("OPENSTACK_AUTH_URL")
	containerName             = "my-container-password-test"
	containerObjects          = []string{
		"1/",
		"1/2/",
		"1/2/3/",
		"1/2/3/4/",
		"1/2/3/hello-world.txt",
		"1/2/hello-world.txt",
		"1/hello-world.txt",
		"2/",
		"2/hello-world.txt",
		"3/",
	}
	minReplicaCount = 0
	maxReplicaCount = 2
)

type templateData struct {
	TestNamespace             string
	DeploymentName            string
	SecretName                string
	TriggerAuthenticationName string
	ScaledObjectName          string
	UserID                    string
	Password                  string
	ProjectID                 string
	AuthURL                   string
	Container                 string
	MinReplicaCount           int
	MaxReplicaCount           int
}

const (
	deploymentTemplate = `apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: test-app
  name: {{.DeploymentName}}
  namespace: {{.TestNamespace}}
spec:
  replicas: 0
  selector:
    matchLabels:
      pod: {{.DeploymentName}}
  template:
    metadata:
      labels:
        pod: {{.DeploymentName}}
    spec:
      containers:
        - name: nginx
          image: 'ghcr.io/nginx/nginx-unprivileged:1.26'
`

	secretTemplate = `
apiVersion: v1
kind: Secret
metadata:
  name: {{.SecretName}}
  namespace: {{.TestNamespace}}
type: Opaque
data:
  userID: {{.UserID}}
  password: {{.Password}}
  projectID: {{.ProjectID}}
  authURL: {{.AuthURL}}
`

	triggerAuthTemplate = `
apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: {{.TriggerAuthenticationName}}
  namespace: {{.TestNamespace}}
spec:
spec:
  secretTargetRef:
  - parameter: userID
    name: {{.SecretName}}
    key: userID
  - parameter: password
    name: {{.SecretName}}
    key: password
  - parameter: projectID
    name: {{.SecretName}}
    key: projectID
  - parameter: authURL
    name: {{.SecretName}}
    key: authURL
`

	scaledObjectTemplate = `apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: {{.ScaledObjectName}}
  namespace: {{.TestNamespace}}
spec:
  scaleTargetRef:
    name: {{.DeploymentName}}
  minReplicaCount: {{.MinReplicaCount}}
  maxReplicaCount: {{.MaxReplicaCount}}
  pollingInterval: 3
  cooldownPeriod:  1
  triggers:
  - type: openstack-swift
    metadata:
      containerName: {{.Container}}
      objectCount: '1'
      activationObjectCount: '5'
    authenticationRef:
      name: {{.TriggerAuthenticationName}}
`
)

func TestScaler(t *testing.T) {
	require.NotEmpty(t, userID, "OPENSTACK_USER_ID env variable is required for Openstack-swift storage test")
	require.NotEmpty(t, password, "OPENSTACK_PASSWORD env variable is required for Openstack-swift storage test")
	require.NotEmpty(t, projectID, "OPENSTACK_PROJECT_ID env variable is required for Openstack-swift storage test")
	require.NotEmpty(t, authURL, "OPENSTACK_AUTH_URL env variable is required for Openstack-swift storage test")

	// Create openstack resources
	client := helper.CreateClient(t, authURL, userID, password, projectID)
	helper.CreateContainer(t, client, containerName)

	kc := GetKubernetesClient(t)
	data, templates := getTemplateData()
	t.Cleanup(func() {
		DeleteKubernetesResources(t, testNamespace, data, templates)
		helper.DeleteContainer(t, client, containerName)
	})

	// Create kubernetes resources
	CreateKubernetesResources(t, kc, testNamespace, data, templates)

	testActivation(t, kc, client)
	testScaleOut(t, kc, client)
	testScaleIn(t, kc, client)
}

func testActivation(t *testing.T, kc *kubernetes.Clientset, client *gophercloud.ServiceClient) {
	t.Log("--- testing activation ---")
	helper.CreateObject(t, client, containerName, containerObjects[0])
	helper.CreateObject(t, client, containerName, containerObjects[1])

	AssertReplicaCountNotChangeDuringTimePeriod(t, kc, deploymentName, testNamespace, minReplicaCount, 60)
}

func testScaleOut(t *testing.T, kc *kubernetes.Clientset, client *gophercloud.ServiceClient) {
	t.Log("--- testing scale out ---")
	for _, object := range containerObjects {
		helper.CreateObject(t, client, containerName, object)
	}

	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, deploymentName, testNamespace, maxReplicaCount, 60, 3),
		"replica count should be %d after 3 minutes", maxReplicaCount)
}

func testScaleIn(t *testing.T, kc *kubernetes.Clientset, client *gophercloud.ServiceClient) {
	t.Log("--- testing scale in ---")
	for _, object := range containerObjects {
		helper.DeleteObject(t, client, containerName, object)
	}
	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, deploymentName, testNamespace, minReplicaCount, 60, 5),
		"replica count should be %d after 5 minutes", minReplicaCount)
}

func getTemplateData() (templateData, []Template) {
	return templateData{
			TestNamespace:             testNamespace,
			DeploymentName:            deploymentName,
			SecretName:                secretName,
			TriggerAuthenticationName: triggerAuthenticationName,
			ScaledObjectName:          scaledObjectName,
			UserID:                    base64.StdEncoding.EncodeToString([]byte(userID)),
			Password:                  base64.StdEncoding.EncodeToString([]byte(password)),
			ProjectID:                 base64.StdEncoding.EncodeToString([]byte(projectID)),
			AuthURL:                   base64.StdEncoding.EncodeToString([]byte(authURL)),
			Container:                 containerName,
			MinReplicaCount:           minReplicaCount,
			MaxReplicaCount:           maxReplicaCount,
		}, []Template{
			{Name: "deploymentTemplate", Config: deploymentTemplate},
			{Name: "secretTemplate", Config: secretTemplate},
			{Name: "triggerAuthTemplate", Config: triggerAuthTemplate},
			{Name: "scaledObjectTemplate", Config: scaledObjectTemplate},
		}
}
